### PR TITLE
tmpfiles: accept additional tmpfiles lines via credential

### PR DESCRIPTION
**Pull request by poettering**

### CODE REVIEW

**Comment by bluca** [comment]
Could you please add a test for this in TEST-54-CREDS?

**Review by bluca — APPROVED** [review]

File: man/systemd-tmpfiles.xml [hunk 1]
```diff
@@ -230,6 +230,29 @@
...
+    <citerefentry><refentrytitle>systemd.exec</refentrytitle><manvolnum>1</manvolnum></citerefentry> for
+    details). The following credentials are used when passed in:</para>
+
+    <variablelist>
+      <varlistentry>
+        <term><literal>tmpfiles.extra</literal></term>
+
+        <listitem><para>The contents of this credential may contain additional lines to operate on. The
+        credential contents should follow the same format as any other <filename>tmpfiles.d/</filename>
+        drop-in.</para></listitem>
```
Comment (yuwata) [inline man/systemd-tmpfiles.xml]: "drop-in" -> "config"?

File: man/systemd-tmpfiles.xml [hunk 2]
```diff
@@ -230,6 +230,29 @@
...
+    <para><command>systemd-tmpfiles</command> supports the service credentials logic as implemented by
+    <varname>LoadCredential=</varname>/<varname>SetCredential=</varname> (see
+    <citerefentry><refentrytitle>systemd.exec</refentrytitle><manvolnum>1</manvolnum></citerefentry> for
+    details). The following credentials are used when passed in:</para>
+
+    <variablelist>
+      <varlistentry>
+        <term><literal>tmpfiles.extra</literal></term>
+
+        <listitem><para>The contents of this credential may contain additional lines to operate on. The
```
Comment (yuwata) [inline man/systemd-tmpfiles.xml]: I think, it is better to mention the priority of the contents. What happens when configs in tmpfiles.d and tmpfiles.extra contradicts?

**Review by yuwata — CHANGES_REQUESTED** [review]

File: man/tmpfiles.d.xml [hunk 3]
```diff
@@ -497,6 +498,16 @@ w- /proc/sys/vm/swappiness - - - - 10</programlisting></para>
       (including newlines and NUL bytes) to files. Note that if this switch is used, the argument is not
       subject to specifier expansion, neither before nor after Base64 decoding.</para>
 
+      <para>If the caret character (<literal>^</literal>) is used, the argument (i.e. 6th) column is takes a
```
Comment (bluca) [inline man/tmpfiles.d.xml]: /is takes/takes/

File: src/tmpfiles/tmpfiles.c [hunk 4]
```diff
@@ -3257,6 +3254,33 @@ static int parse_line(
                 }
         }
 
+        if (from_cred) {
+                if (!i.argument)
+                        return log_syntax(NULL, LOG_ERR, fname, line, SYNTHETIC_ERRNO(EINVAL), "Reading from credential requested, but not credential name specified.");
```
Comment (yuwata) [inline src/tmpfiles/tmpfiles.c]: not -> no

File: src/tmpfiles/tmpfiles.c [hunk 5]
```diff
@@ -3257,6 +3254,33 @@ static int parse_line(
...
+                if (!credential_name_valid(i.argument))
+                        return log_syntax(NULL, LOG_ERR, fname, line, SYNTHETIC_ERRNO(EINVAL), "Credential name not valid: %s", i.argument);
+
+                r = read_credential(i.argument, &i.binary_argument, &i.binary_argument_size);
+                if (IN_SET(r, -ENXIO, -ENOENT)) {
```
Comment (yuwata) [inline src/tmpfiles/tmpfiles.c:3264]: How about the other error case?

**Review by yuwata — CHANGES_REQUESTED** [review]

**Comment by yuwata** [comment]
LGTM. But please rebase.

**Comment by yuwata** [comment]
Lint Code Base CI does not like the test code:
```
In /github/workspace/test/TEST-54-CREDS/test.sh line 7:
NSPAWN_ARGUMENTS="${NSPAWN_ARGUMENTS:-} "\
                                         ^-- SC2140 (warning): Word is of the form "A"B"C" (B indicated). Did you mean "ABC" or "A\"B\"C"?
```
How about the following?
```sh
NSPAWN_CREDS=(
    "--set-credential=mynspawncredential:strangevalue"
)
NSPAWN_ARGUMENTS="${NSPAWN_ARGUMENTS:-} ${NSPAWN_CREDS[*]}"

QEMU_CREDS=(
    "-fw_cfg name=opt/io.systemd.credentials/myqemucredential,string=othervalue"
    "-smbios type=11,value=io.systemd.credential:smbioscredential=magicdata"
    "-smbios type=11,value=io.systemd.credential.binary:binarysmbioscredential=bWFnaWNiaW5hcnlkYXRh"
    "-smbios type=11,value=io.systemd.credential.binary:sysusers.extra=dSBjcmVkdGVzdHVzZXIK"
    "-smbios type=11,value=io.systemd.credential.binary:tmpfiles.extra=ZiAvdG1wL3NvdXJjZWRmcm9tY3JlZGVudGlhbCAtIC0gLSAtIHRtcGZpbGVzc2VjcmV0Cg=="
)
QEMU_OPTIONS="${QEMU_OPTIONS:-} ${QEMU_CREDS[*]}"

KERNEL_CREDS=(
    "systemd.set_credential=kernelcmdlinecred:uff"
    "systemd.set_credential=sysctl.extra:kernel.domainname=sysctltest"
    "systemd.set_credential=login.motd:hello"
    "systemd.set_credential=login.issue:welcome"
    "rd.systemd.import_credentials=no"
)
KERNEL_APPEND="${KERNEL_APPEND:-} ${KERNEL_CREDS[*]}"
```

**Comment by poettering** [comment]
> How about the following?

Excellent! That works and looks pretty! Thank you!

Will rebase with that and push a new version shortly. PTAL

**Review by yuwata — APPROVED** [review]

**Comment by bluca** [comment]
Could this be related? Any idea @mrc0mmand ?

```
00:48:39 FAILED TASKS:
00:48:39 -------------
00:48:39 TEST-35-LOGIN
```

**Comment by mrc0mmand** [comment]
> Could this be related? Any idea @mrc0mmand ?
> 
> ```
> 00:48:39 FAILED TASKS:
> 00:48:39 -------------
> 00:48:39 TEST-35-LOGIN
> ```

Nah, that's #24040.

**Pull request merged**